IONFileTrasnsferLib/Delegates/IONFLTRUploadDelegate.swift

```diff
@@ -0,0 +1,96 @@
+    ///
+    /// - Parameters:
+    ///   - publisher: The publisher used to send progress and success updates.
+    ///   - disableRedirects: A flag indicating whether HTTP redirects should be disabled.
+    init(publisher: IONFLTRPublisher, disableRedirects: Bool) {
+        self.publisher = publisher
+        self.disableRedirects = disableRedirects
+    }
+}
+
+extension IONFLTRUploadDelegate: URLSessionTaskDelegate {
+    
+    /// Reports the progress of an upload task.
+    ///
+    /// - Parameters:
+    ///   - session: The `URLSession` containing the upload task.
+    ///   - task: The `URLSessionTask` reporting progress.
+    ///   - bytesSent: The number of bytes sent since the last call.
+    ///   - totalBytesSent: The total number of bytes sent so far.
+    ///   - totalBytesExpectedToSend: The total number of bytes expected to be sent.
+    func urlSession(_ session: URLSession, task: URLSessionTask, didSendBodyData bytesSent: Int64, totalBytesSent: Int64, totalBytesExpectedToSend: Int64) {
+        self.totalBytesSent = totalBytesSent
+        publisher.sendProgress(totalBytesSent, totalBytesExpected: totalBytesExpectedToSend)
+    }
+    
+    /// Handles the completion of an upload task.
+    ///
+    /// - Parameters:
+    ///   - session: The `URLSession` containing the task.
+    ///   - task: The `URLSessionTask` that completed.
+    ///   - error: The error that occurred, if any.
+    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: (any Error)?) {
+        if let error = error {
+            publisher.sendFailure(error)
+        }
+    }
+    
+    /// Handles HTTP redirection for a task.
+    ///
+    /// - Parameters:
+    ///   - session: The `URLSession` containing the task.
+    ///   - task: The `URLSessionTask` that will be redirected.
+    ///   - response: The `HTTPURLResponse` that caused the redirection.
+    ///   - request: The new `URLRequest` to redirect to.
+    ///   - completionHandler: A closure to execute with the new request or `nil` to refuse the redirect.
+    func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse, newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Void) {
+        if disableRedirects {
+            completionHandler(nil)
+        } else {
+            completionHandler(request)
+        }
+    }
+}
+
+extension IONFLTRUploadDelegate: URLSessionDataDelegate {
+    
+    /// Handles the receipt of data during an upload task.
+    ///
+    /// - Parameters:
+    ///   - session: The `URLSession` containing the data task.
+    ///   - dataTask: The `URLSessionDataTask` that received the data.
+    ///   - data: The data received from the server.
+    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {
+        let response = dataTask.response as? HTTPURLResponse
```

Comment:
Same comment as above, should we use `guard` here and return an error in case `dataTask.response as? HTTPURLResponse` fails?

Reply:
Some something like:

```
guard let response = dataTask.response as? HTTPURLResponse else {
    publisher.sendFailure(someExceptionorError)
    return
}
```

Reply:
Same as above.

Reply:
Yeah that makes sense, I'll resolve this conversation.